# Data Cleaning with Pandas

In [370]:
import pandas as pd

## Scenario

As data scientists, we want to build a model to predict the sale price of a house in Seattle in 2019, based on its square footage. We know that the King County Department of Assessments has comprehensive data available on real property sales in the Seattle area. We need to prepare the data.

### First, get the data!

When working on a project involving data that can fit on our computer, we store it in a `data` directory.

```bash
cd <project_directory>  # example: cd ~/flatiron_ds/pandas-3
mkdir data
cd data
```

Note that `<project_directory>` in angle brackets is a _placeholder_. You should type the path to the actual location on your computer where you're working on this project. Do not literally type `<project_directory>` and _do not type the angle brackets_. You can see an example in the _comment_ to the right of the command above.

Now, we'll need to download the two data files that we need. We can do this at the command line:

```bash
wget https://aqua.kingcounty.gov/extranet/assessor/Real%20Property%20Sales.zip
wget https://aqua.kingcounty.gov/extranet/assessor/Residential%20Building.zip
```

*Note:* If you do not have the `wget` command yet, you can install it: `brew install wget`.

Note that `%20` in a URL translates into a space. Even though you will *never put spaces in filenames*, you may need to deal with spaces that _other_ people have used in filenames.

There are two ways to handle the spaces in these filenames when referencing them at the command line.

#### 1. You can _escape_ the spaces by putting a backslash (`\`, remember _backslash is next to backspace_) before each one:

`unzip Real\ Property\ Sales.zip`

This is what happens if you tab-complete the filename in the terminal. Tab completion is your friend!

#### 2. You can put the entire filename in quotes:

`unzip "Real Property Sales.zip"`

Try unzipping these files with the `unzip` command. The `unzip` command takes one argument, the name of the tile that you want to unzip.


You can use tab completion (press the `tab` key after the first three letters) to fill in the names, including spaces. This will 

In [371]:
sales_df = pd.read_csv('data/EXTR_RPSale.csv') #mixed types means there are int and str, etc. so it stores everything as string

/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Seeing pink? Warnings are useful!

Note the warning above: `DtypeWarning: Columns (1, 2) have mixed types.` Because we start with an index of zero, the columns that we're being warned about are actually the _second_ and _third_ columns, `sales_df['Major']` and `sales_df['Minor']`.

In [372]:
sales_df.head().T

0  \
ExciseTaxNbr                                                  2687551   
Major                                                          138860   
Minor                                                             110   
DocumentDate                                               08/21/2014   
SalePrice                                                      245000   
RecordingNbr                                           20140828001436   
Volume                                                                  
Page                                                                    
PlatNbr                                                                 
PlatType                                                                
PlatLot                                                                 
PlatBlock                                                               
SellerName          WENKLE NOEL SMITH -TRUSTEE                    ...   
BuyerName           ALEXANDER APRIL                               ...   
PropertyType                                                        3   
PrincipalUse                                                        6   
SaleInstrument                                                      3   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                          1   
PropertyClass                                                       8   
SaleWarning                                                             

                                                                    1  \
ExciseTaxNbr                                                  1235111   
Major                                                          664885   
Minor                                                              40   
DocumentDate                                               07/09/1991   
SalePrice                                                           0   
RecordingNbr                                           199203161090     
Volume                                                            071   
Page                                                              001   
PlatNbr                                                        664885   
PlatType                                                            C   
PlatLot                                                B102             
PlatBlock                                                               
SellerName          HERRMAN JEFFREY A+PAULINE A                   ...   
BuyerName           HERRMAN PAULINE A TTEE                        ...   
PropertyType                                                        3   
PrincipalUse                                                        0   
SaleInstrument                                                     26   
AFForestLand                                                        N   
AFCurrentUseLand                                                    N   
AFNonProfitUse                                                      N   
AFHistoricProperty                                                  N   
SaleReason                                                         18   
PropertyClass                                                       3   
SaleWarning                                                        11   

                                                                    2  \
ExciseTaxNbr                                                  2704079   
Major                                                          423943   
Minor                                                              50   
DocumentDate                                               10/11/2014   
SalePrice                                          

### Data overload?

That's a lot of columns. We're only interested in identifying the date, sale price, and square footage of each specific property. What can we do?

In [373]:
sales_df = sales_df[['Major', 'Minor', 'DocumentDate', 'SalePrice']]

In [374]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.5+ MB


In [375]:
bldg_df = pd.read_csv('data/Residential Building.zip')

/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3020: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [376]:
bldg_df

Major  Minor  BldgNbr  NbrLivingUnits                        Address  \
0         4300    167        1               1      15223   40TH AVE S  98188   
1         4610    399        1               1    4431   FERNCROFT RD   98040   
2         4610    503        1               1    4516   FERNCROFT RD   98040   
3         4610    505        1               1    4538   FERNCROFT RD   98040   
4         4900     56        1               1      3015  SW 105TH ST   98146   
5         4950     60        1               1     2508   VASHON CT NE  98059   
6         5030    300        1               1     26115   201ST PL SE  98042   
7         5030    350        1               1     20101  SE 260TH PL   98042   
8         5100    160        1               1      2115  SW 305TH ST   98023   
9         5200     21        1               1            1212 A  6TH AVE N     
10        5350    300        1               1         21366  SE 289TH WAY      
11        5350    510        1               1         21371  SE 289TH WAY      
12        5350    790        1               1          21561  SE 290TH PL      
13        5600    145        1               1   5722  SW WINTHROP ST   98116   
14        5600    230        1               1   5717  SW WINTHROP ST   98116   
15        5900    125        1               1   5353  SW ADMIRAL WAY   98116   
16        5900    220        1               1      2721   53RD AVE SW  98116   
17        5900    250        1               1   5509  SW CAMPBELL PL   98116   
18        5900    345        1               1   5557  SW CAMPBELL PL   98116   
19        5900    420        1               1   5344  SW ADMIRAL WAY   98116   
20        5900    535        1               1   5514  SW CAMPBELL PL   98116   
21        5950    160        1               1       3719  S 270TH ST   98032   
22        6500    140        1               1      6521   36TH AVE SW  98126   
23        6500    190        1               1      6546   37TH AVE SW  98126   
24        6500    230        1               1      6506   37TH AVE SW  98126   
25        6500    235        1               1      6502   37TH AVE SW  98126   
26        6500    330        1               1      6506   38TH AVE SW  98126   
27        6600     75        1               1      5952   38TH AVE SW  98126   
28        6600    165        1               1      5937   38TH AVE SW  98126   
29        6600    340        1               1      6026   39TH AVE SW  98136   
...        ...    ...      ...             ...                            ...   
511382  394190     72        1               1       9532   6TH AVE NW  98117   
511383  394190     82        1               1       9521   4TH AVE NW  98117   
511384  394550     70        1               1      2143  SW 317TH PL   98023   
511385  394690     60        1               1        312  NE 88TH ST   98115   
511386  395390     80        1               1     316  S DIVISION ST   98002   
511387  395550     10        1               1    11458   108TH AVE NE  98033   
511388  395590     70        1               1        2324  SE 2ND PL   98056   
511389  395590    170        1               1      258   BLAINE CT SE  98056   
511390  395590    340        1               1        2100  SE 2ND PL   98056   
511391  395590    380        1               1      255   BLAINE DR SE  98056   
511392  395590    430        1               1      265   BLAINE DR SE  98056   
511393  395590    520        1               1        2207  SE 2ND PL   98056   
511394  395590    960        1               1        2235  SE 3RD ST   98056   
511395  395590   1150        1               1    318   EDMONDS AVE SE  98056   
511396  395590   1260        1               1    265   EDMONDS AVE SE  98056   
511397  395590   1280        1               1    277   EDMONDS AVE SE  98056   
511398  395610     20        1               1  1310   LA FOREST DR SE  98045   
511399  395610    100        1          

### Another warning! Which column has index 11?

In [377]:
bldg_df.columns[11]

'ZipCode'

`ZipCode` seems like a potentially useful column. We'll need it to determine which house sales took place in Seattle.

In [378]:
bldg_df.head().T

0                            1  \
Major                                    4300                         4610   
Minor                                     167                          399   
BldgNbr                                     1                            1   
NbrLivingUnits                              1                            1   
Address             15223   40TH AVE S  98188  4431   FERNCROFT RD   98040   
BuildingNumber                          15223                        4431    
Fraction                                                                     
DirectionPrefix                                                              
StreetName          40TH                         FERNCROFT                   
StreetType                             AVE                          RD       
DirectionSuffix                            S                                 
ZipCode                                 98188                        98040   
Stories                                     1                            2   
BldgGrade                                   6                            9   
BldgGradeVar                                0                            0   
SqFt1stFloor                              990                          980   
SqFtHalfFloor                               0                            0   
SqFt2ndFloor                                0                         1140   
SqFtUpperFloor                              0                            0   
SqFtUnfinFull                               0                            0   
SqFtUnfinHalf                               0                            0   
SqFtTotLiving                             990                         2950   
SqFtTotBasement                           990                          940   
SqFtFinBasement                             0                          830   
FinBasementGrade                            0                            9   
SqFtGarageBasement                          0                            0   
SqFtGarageAttached                          0                          480   
DaylightBasement                                                         Y   
SqFtOpenPorch                              80                            0   
SqFtEnclosedPorch                           0                            0   
SqFtDeck                                    0                          600   
HeatSystem                                  6                            5   
HeatSource                                  1                            2   
BrickStone                                  0                            0   
ViewUtilization                                                          N   
Bedrooms                                    2                            3   
BathHalfCount                               0                            1   
Bath3qtrCount                               0                            0   
BathFullCount                               1                            2   
FpSingleStory                               0                            0   
FpMultiStory                                0                            0   
FpFreestanding                              0                            2   
FpAdditional                                0                            0   
YrBuilt                                  1922                         1978   
YrRenovated                                 0                            0   
PcntComplete                                0                            0   
Obsolescence                                0                            0   
PcntNetCondition                            0                            0   
Condition                                   3                            4   
AddnlCost                                   0                            0   

                                              2                  

### So many features!

As data scientists, we should be _very_ cautious about discarding potentially useful data. But, today, we're interested in _only_ the total square footage of each property. What can we do?


In [379]:
bldg_df = bldg_df[['Major', 'Minor', 'SqFtTotLiving', 'ZipCode']]

In [380]:
bldg_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511412 entries, 0 to 511411
Data columns (total 4 columns):
Major            511412 non-null int64
Minor            511412 non-null int64
SqFtTotLiving    511412 non-null int64
ZipCode          468372 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB


In [381]:
#sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])
#Major section is object on one data set and an int on the other

### Error!

Why are we seeing an error when we try to join the dataframes?

<table>
    <tr>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2013160 entries, 0 to 2013159
Data columns (total 4 columns):
Major           object
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: int64(1), object(3)
memory usage: 61.4+ MB</pre></td>
        <td style="text-align:left"><pre>
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 511359 entries, 0 to 511358
Data columns (total 4 columns):
Major            511359 non-null int64
Minor            511359 non-null int64
SqFtTotLiving    511359 non-null int64
ZipCode          468345 non-null object
dtypes: int64(3), object(1)
memory usage: 15.6+ MB
</pre></td>
    </tr>
</table>

Review the error message in light of the above:

* `ValueError: You are trying to merge on object and int64 columns.`

In [382]:
#sales_df['Major'] = pd.to_numeric(sales_df['Major'])

### Error!

Note the useful error message above:

`ValueError: Unable to parse string "      " at position 936643`

In this case, we want to treat non-numeric values as missing values. Let's see if there's a way to change how the `pd.to_numeric` function handles errors.

In [383]:
# The single question mark means "show me the docstring"
pd.to_numeric?


Here's the part that we're looking for:
```
errors : {'ignore', 'raise', 'coerce'}, default 'raise'
    - If 'raise', then invalid parsing will raise an exception
    - If 'coerce', then invalid parsing will be set as NaN
    - If 'ignore', then invalid parsing will return the input
```

Let's try setting the `errors` parameter to `'coerce'`.

In [384]:
sales_df['Major'] = pd.to_numeric(sales_df['Major'], errors='coerce')

Did it work?

In [385]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           object
DocumentDate    object
SalePrice       int64
dtypes: float64(1), int64(1), object(2)
memory usage: 61.5+ MB


It worked! Let's do the same thing with the `Minor` parcel number.

In [386]:
sales_df['Minor'] = pd.to_numeric(sales_df['Minor'], errors='coerce')

In [387]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2014336 entries, 0 to 2014335
Data columns (total 4 columns):
Major           float64
Minor           float64
DocumentDate    object
SalePrice       int64
dtypes: float64(2), int64(1), object(1)
memory usage: 61.5+ MB


Now, let's try our join again.

In [388]:
sales_data = pd.merge(sales_df, bldg_df, on=['Major', 'Minor'])

In [389]:
sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

In [390]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1437522 entries, 0 to 1437521
Data columns (total 6 columns):
Major            1437522 non-null float64
Minor            1437522 non-null float64
DocumentDate     1437522 non-null object
SalePrice        1437522 non-null int64
SqFtTotLiving    1437522 non-null int64
ZipCode          1322148 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 76.8+ MB


We can see right away that we're missing zip codes for many of the sales transactions. (1321536 non-null entries for ZipCode is fewer than the 1436772 entries in the dataframe.) 

In [391]:
sales_data.loc[sales_data['ZipCode'].isna()].head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
99   858140.0  376.0   05/22/2012          0            900     NaN
100  858140.0  376.0   11/28/2017          0            900     NaN
152  720319.0  520.0   11/20/2013     699950           2840     NaN
153  720319.0  520.0   09/16/2013          0           2840     NaN
163  894677.0  240.0   12/21/2016     818161           2450     NaN

Because we are interested in finding houses in Seattle zip codes, we will need to drop the rows with missing zip codes.

In [392]:
sales_data = sales_data.loc[~sales_data['ZipCode'].isna(), :]
#.isna() function changes NaN to true or false; 1 or 0 

sales_data.head()

Major  Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
0  138860.0  110.0   08/21/2014     245000           1490   98002
1  138860.0  110.0   06/12/1989     109300           1490   98002
2  138860.0  110.0   01/16/2005      14684           1490   98002
3  138860.0  110.0   06/08/2005          0           1490   98002
4  423943.0   50.0   10/11/2014          0            960   98092

# Your turn: Data Cleaning with Pandas

### 1. Investigate and drop rows with invalid values in the SalePrice and SqFtTotLiving columns.

Use multiple notebook cells to accomplish this! Press `[esc]` then `B` to create a new cell below the current cell. Press `[return]` to start typing in the new cell.

In [393]:
sales_data_i = sales_data[['SalePrice', 'SqFtTotLiving', 'DocumentDate']]
sales_data_i.head()

SalePrice  SqFtTotLiving DocumentDate
0     245000           1490   08/21/2014
1     109300           1490   06/12/1989
2      14684           1490   01/16/2005
3          0           1490   06/08/2005
4          0            960   10/11/2014

In [394]:
sales_data_i=sales_data_i.loc[sales_data_i['SalePrice']!=0] #you have to set it to itself in order to overwrite
sales_data_i.head()

SalePrice  SqFtTotLiving DocumentDate
0     245000           1490   08/21/2014
1     109300           1490   06/12/1989
2      14684           1490   01/16/2005
6      96000            960   07/15/1999
7     127500            960   01/08/2001

### 2. Investigate and handle non-numeric ZipCode values

Can you find a way to shorten ZIP+4 codes to the first five digits?

What's the right thing to do with missing values?

In [395]:
# Read the error message and decide how to fix it.
# Note: using errors='coerce' is the *wrong* choice in this case.
def is_integer(x):
    try:
        _ = int(x)
    except ValueError:
        return False
    return True

sales_data.loc[sales_data['ZipCode'].apply(is_integer) == False, 'ZipCode'].str[:5].head()

10168    98042
10169    98042
10170    98042
10171    98042
10172    98042
Name: ZipCode, dtype: object

### 3. Add a column for PricePerSqFt



In [396]:
sales_data_i.head()

SalePrice  SqFtTotLiving DocumentDate
0     245000           1490   08/21/2014
1     109300           1490   06/12/1989
2      14684           1490   01/16/2005
6      96000            960   07/15/1999
7     127500            960   01/08/2001

In [397]:
PricePerSqFt = sales_data_i['SalePrice'] / sales_data_i['SqFtTotLiving']
PricePerSqFt.head()

0    164.429530
1     73.355705
2      9.855034
6    100.000000
7    132.812500
dtype: float64

In [398]:
sales_data_i['PricePerSqFt'] = PricePerSqFt.values

In [419]:
sales_data_i_2019 = sales_data_i.assign(PricePerSqFt=PricePerSqFt.values)

In [420]:
sales_data_i_2019.head()

SalePrice  SqFtTotLiving DocumentDate  PricePerSqFt
0     245000           1490   2014-08-21    164.429530
1     109300           1490   1989-06-12     73.355705
2      14684           1490   2005-01-16      9.855034
6      96000            960   1999-07-15    100.000000
7     127500            960   2001-01-08    132.812500

In [401]:
sales_data_i.round(2).head()

SalePrice  SqFtTotLiving DocumentDate  PricePerSqFt
0     245000           1490   08/21/2014        164.43
1     109300           1490   06/12/1989         73.36
2      14684           1490   01/16/2005          9.86
6      96000            960   07/15/1999        100.00
7     127500            960   01/08/2001        132.81

In [402]:
sales_data_i = sales_data_i.round(2)

In [403]:
sales_data_i.head()

SalePrice  SqFtTotLiving DocumentDate  PricePerSqFt
0     245000           1490   08/21/2014        164.43
1     109300           1490   06/12/1989         73.36
2      14684           1490   01/16/2005          9.86
6      96000            960   07/15/1999        100.00
7     127500            960   01/08/2001        132.81

### 4. Subset the data to 2019 sales only.

We can assume that the DocumentDate is approximately the sale date.

In [404]:
sales_data_i[['DocumentDate']].head()

DocumentDate
0   08/21/2014
1   06/12/1989
2   01/16/2005
6   07/15/1999
7   01/08/2001

In [405]:
type(sales_data_i['DocumentDate'])

pandas.core.series.Series

In [406]:
sales_data_i.head()

SalePrice  SqFtTotLiving DocumentDate  PricePerSqFt
0     245000           1490   08/21/2014        164.43
1     109300           1490   06/12/1989         73.36
2      14684           1490   01/16/2005          9.86
6      96000            960   07/15/1999        100.00
7     127500            960   01/08/2001        132.81

In [407]:
sales_date = sales_data['DocumentDate']
sales_date.head()

0    08/21/2014
1    06/12/1989
2    01/16/2005
3    06/08/2005
4    10/11/2014
Name: DocumentDate, dtype: object

In [430]:
sales_data_i_2019.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 867416 entries, 0 to 1437520
Data columns (total 4 columns):
SalePrice        867416 non-null int64
SqFtTotLiving    867416 non-null int64
DocumentDate     867416 non-null datetime64[ns]
PricePerSqFt     867416 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(2)
memory usage: 33.1 MB


In [431]:
sales_data_i_2019['DocumentDate'] = pd.to_datetime(sales_data_i['DocumentDate'])
#to_datetime converts an object to a date format

In [433]:
#sales_data[sales_date.to_datetime.year == 2019]
sales_data_i_2019=sales_data_i_2019[sales_data_i_2019["DocumentDate"].dt.year == 2019]
sales_data_i_2019.head()


SalePrice  SqFtTotLiving DocumentDate  PricePerSqFt
605      390000           1780   2019-03-06    219.101124
651      923000           2550   2019-03-13    361.960784
868      935000           2460   2019-02-06    380.081301
1063     890000           2380   2019-02-15    373.949580
1341     380000           1750   2019-03-04    217.142857

### 5. Subset the data to zip codes within the City of Seattle.

You'll need to find a list of Seattle zip codes!

In [411]:
sales_data_i_2019=sales_data_i_2019.loc[sales_data_i_2019['SalePrice']!=0] 


In [418]:
sales_data_i_2019 .head()

Major   Minor DocumentDate  SalePrice  SqFtTotLiving ZipCode
441  661000.0  1020.0   2002-02-01      80694           1660   98103
443  661000.0  1020.0   1994-05-04     185000           1660   98103
579  503530.0   206.0   1993-08-25     185000           1350   98199
582  503530.0   206.0   2007-10-11     625000           1350   98199
778  890100.0  1043.0   1996-04-16     149000           1750   98125

In [413]:
seattle_zipcodes = [98101, 98102, 98103, 98104, 98105, 98106, 98107, 98108, 98109, 98112, 98115, 98116, 98117, 98118, 98119, 98121, 98122, 98125, 98126, 98133, 98134, 98136, 98144, 98146, 98154, 98164, 98174, 98177, 98178, 98195, 98199]

In [414]:
seattle_zipcodes = list(seattle_zipcodes)
seattle_zipcodes

[98101,
 98102,
 98103,
 98104,
 98105,
 98106,
 98107,
 98108,
 98109,
 98112,
 98115,
 98116,
 98117,
 98118,
 98119,
 98121,
 98122,
 98125,
 98126,
 98133,
 98134,
 98136,
 98144,
 98146,
 98154,
 98164,
 98174,
 98177,
 98178,
 98195,
 98199]

In [415]:
def in_seattle(x):
    seattle_zipcodes
    return x in seattle_zipcodes
sales_data_i_2019 = sales_data_i_2019.loc[sales_data_i_2019['ZipCode'].apply(in_seattle) == True, :]
    



In [437]:
sales_data_i_2019.head()

SalePrice  SqFtTotLiving DocumentDate  PricePerSqFt
605      390000           1780   2019-03-06    219.101124
651      923000           2550   2019-03-13    361.960784
868      935000           2460   2019-02-06    380.081301
1063     890000           2380   2019-02-15    373.949580
1341     380000           1750   2019-03-04    217.142857

### 6. What is the mean price per square foot for a house sold in Seattle in 2019?

Don't just type the answer. Type code that generates the answer as output!

In [435]:
sales_data_i_2019['PricePerSqFt'].mean()

401.4669645065364